source/browseMode.py

```diff
@@ -2001,6 +2040,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
 				yield TextInfoQuickNavItem("notLinkBlock", self, textRange)
 			item1=item2
 
+	STYLE_ATTRIBUTES = frozenset([
+		'background-color',
+		'color',
+		'font-family,',
+		'font-size',
+		'bold',
+		'italic',
+		'marked',
+		'strikethrough',
+		'text-line-through-style',
+		'underline',
+		'text-underline-style',
+	])
+
+	def _extractStyles(
+			self,
+			info: textInfos.TextInfo,
+	) -> "textInfos.TextInfo.TextWithFieldsT":
+		"""
+		This function calls TextInfo.getTextWithFields(), and then processes fiields in the following way:
+		1. Highlighted(marked) text is currently reported as Role.MARKED_CONTENT, and not formatChange.
+		For ease of further handling we create a new boolean format field "marked"
+		and set its value according to presence of Role.MARKED_CONTENT.
+		2. Then we drop all control fields, leaving only formatChange fields and text.
+		"""
+		stack = [{}]
+		result = []
+		for field in info.getTextWithFields():
+			if isinstance(field, textInfos.FieldCommand):
+				if field.command == 'controlStart':
+					style = {**stack[-1]}
+					if field.field['role'] == controlTypes.Role.MARKED_CONTENT:
+						style['marked'] = True
+					stack.append(style)
+				elif field.command == 'controlEnd':
+					del stack[-1]
+				elif field.command == 'formatChange':
+					field.field = {
+						k: v
+						for k, v in {**field.field, **stack[-1]}.items()
+						if k in self.STYLE_ATTRIBUTES
+					}
+					result.append(field)
+				else:
+					raise RuntimeError()
+			elif isinstance(field, str):
+				result.append(field)
+			else:
+				raise RuntimeError
+		return result
+
+	def _iterTextStyle(
+			self,
+			kind: str,
+			direction: str = "next",
+			pos: textInfos.TextInfo = None
+	) -> Generator[TextInfoQuickNavItem, None, None]:
+		sameStyle = kind == 'sameStyle'
+		initialTextInfo = pos.copy()
+		initialTextInfo.collapse()
+		result = initialTextInfo.move(textInfos.UNIT_CHARACTER, 1, endPoint='end')
+		if result == 0:
+			result = initialTextInfo.move(textInfos.UNIT_CHARACTER, -1, endPoint='start')
+			if result == 0:
+				# Translators: Error message for same/different style QuickNav command
+				ui.message(_("Cannot determine current style"))

```

Comment:
Is it an expected error? If yes, could you describe a case?
* If the error is expected in some case, there is no point in raising an error. Just return if needed, and log a debugWarning if it seems of interest, so that we can see that we are in a corner case.
* If the error is not expected: there is no point in using a `ui.message`. The user will already have no message, when expecting one, and there will be an error in the log.

Reply:
If I simply return from this generator, then `_quickNavScript` function that calls the generator would report "no next same style text" - which is wrong.
The error is not really expected, however I think this is a good style to check for potential problems in code. Like there are many different TextInfo implementations and all of them have slightly different behaviors. So it doesn't hurt to speak a message if one of sanity checks doesn't pass. However if you strongly feel that we should raise Exception without a message - I can remove it.

Reply:
OK. Thanks for the clarification.

This pattern with a UI message and an error is not so common in NVDA. But if NV Access is OK, I have no objection.

Generally It may have been still better to call the `ui.message` in the calling function in a try block. But in the present case, it does not seem applicable due to code being factorized.